File: session07/BinaryTree.java
public class BinaryTree {
	private static class Node {
		private int val;
		Node left, right;

		Node(int v) { val = v; left = null; right = null; }

		public void preorder(StringBuilder b) {
			b.append(val);			
			if (left != null)
				left.preorder();
			if (right != null)
				right.preorder(b);
		}

		public void postorder(StringBuilder b) {
			if (left != null)
				left.postorder();
			if (right != null)
				right.postorder(b);
			b.append(val);			
		}

		public void inorder(StringBuilder b) { // x + y
			if (left != null)
				left.inorder(b);
			b.append(val);
			if (right != null)
				right.inorder(b);
			
		}
	}

	private Node root;
	public BinaryTree() {
		root = null;
	}

	public void add(int v) {
		if (root == null) { // O(1)
			root = new Node(v);
			return;
		}
		Node p = root;
		while(true)
			if (v > p.val) {
				if (p.right == null) {
					p.right = new Node(v);
					return;
				}
				p = p.right;
			} else {
				if (p.left == null) {
					p.left = new Node(v);
					return;
				}
				p = p.left;
			}

	}
	
	public static void main(String[] args) {
		BinaryTree b = new BinaryTree();
		b.add(3);
	}

File: session07/BinaryTree.cc
#include <iostream>
using namespace std;

class BinaryTree {
private:
  class Node {
  public:
    int val;
    Node* left;				// int* a,b; a is pointer to int, b is just int
    Node* right;
    Node(int v, Node* L, Node* R) : val(v), left(L), right(R) {}


		void inorder(ostream& s) { // first do left, then me, then right
			if (left != nullptr)
				left->inorder(s);
			s << val;
			if (right != nullptr)
				right->inorder(s);
		}

		void preorder(ostream& s) { // first do me, then left right, then right child
			s << val;
			if (left != nullptr)
				left->inorder(s);
			if (right != nullptr)
				right->inorder(s);
		}

		void postorder(ostream& s) { // first do left, then right, then me
			if (left != nullptr)
				left->inorder(s);
			if (right != nullptr)
				right->inorder(s);
			s << val;
		}
		void addRecursive(int v) {
			if (v > val) {
				right->addRecursive(v);
			} else {
				left->addRecursive(v);
			}
		}

  };
  Node * root;
public:
	BinaryTree() : root(nullptr) {}
	~BinaryTree() {}	
  void add(int v) {  // O(n)
    if (root == nullptr) {
      root = new Node(v, nullptr, nullptr);
      return;
    }
		Node* p = root;
		while(true) {
			if (v > p->val) {
				if (p->right == nullptr) {
					p->right = new Node(v, nullptr, nullptr);
					return;
				} else
					p = p->right;
			} else {
				if (p->left == nullptr) {
					p->left = new Node(v, nullptr, nullptr);
					return;
				} else
					p = p->left;
			}
		}
	}
public:
	void addRecursive(int v) {
    if (root == nullptr) {
      root = new Node(v, nullptr, nullptr);
      return;
    }
		root->addRecursive(v);
	}


  void preorder(ostream& s) {
		if (root == nullptr)
			return;
		root->preorder(s);
	}

	void postorder(ostream& s) {
		if (root == nullptr)
			return;
		root->postorder(s);
	}
	void inorder(ostream& s) {
		if (root == nullptr)
			return;
		root->inorder(s);
	}


	friend ostream& operator <<(ostream& s, const BinaryTree& tree) {
		tree.root->inorder(s);
		//tree.root->postorder(s);
		//tree.root->preorder(s);
    return s;
	}
};

int main(){
  BinaryTree b;
	b.add(5);
	for (int i = 0; i < 10; i++)
		b.add(i);
	cout << b;
}
